src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Activity.cs

```diff
@@ -125,6 +126,74 @@ public async Task<IEnumerable<InstanceMetadata>> UpdateInstanceWatermarkV2Async(
+    public async Task<UpdateInstanceResponse> UpdateInstanceBlobsV3Async(
+        [ActivityTrigger] UpdateInstanceBlobArgumentsV2 arguments,
+        ILogger logger)
+    {
+        EnsureArg.IsNotNull(arguments, nameof(arguments));
+        EnsureArg.IsNotNull(arguments.ChangeDataset, nameof(arguments.ChangeDataset));
+        EnsureArg.IsNotNull(arguments.InstanceMetadataList, nameof(arguments.InstanceMetadataList));
+        EnsureArg.IsNotNull(arguments.Partition, nameof(arguments.Partition));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        DicomDataset datasetToUpdate = GetDeserialzedDataset(arguments.ChangeDataset);
+
+        logger.LogInformation("Beginning to update all instance blobs, Total count {TotalCount}", arguments.InstanceMetadataList.Count);
+
+        var updatedInstances = new ConcurrentBag<InstanceMetadata>();
+        var errors = new ConcurrentBag<string>();
+
+        await Parallel.ForEachAsync(
+            arguments.InstanceMetadataList,
+            new ParallelOptions
+            {
+                CancellationToken = default,
+                MaxDegreeOfParallelism = _options.MaxParallelThreads,
+            },
+            async (instance, token) =>
+            {
+                try
+                {
+                    FileProperties fileProperties = await _updateInstanceService.UpdateInstanceBlobAsync(instance, datasetToUpdate, arguments.Partition, token);
+                    updatedInstances.Add(
+                        new InstanceMetadata(
+                            instance.VersionedInstanceIdentifier,
+                            new InstanceProperties
+                            {
+                                FileProperties = fileProperties,
+                                NewVersion = instance.InstanceProperties.NewVersion,
+                                OriginalVersion = instance.InstanceProperties.OriginalVersion
+                            }));
+                }
+                catch (DataStoreRequestFailedException ex)
+                {
+                    logger.LogInformation("Failed to update instance with watermark {Watermark}", instance.VersionedInstanceIdentifier.Version);
+                    errors.Add($"{ex.Message}. {ToInstanceString(instance.VersionedInstanceIdentifier)}");
+                }
+                catch (DataStoreException)
+                {
+                    logger.LogInformation("Failed to update instance with watermark {Watermark}", instance.VersionedInstanceIdentifier.Version);
+                    errors.Add($"Failed to update instance. {ToInstanceString(instance.VersionedInstanceIdentifier)}");
+                }
+            });
+
+        logger.LogInformation("Completed updating all instance blobs. Total instace count {TotalCount}. Total Failed {FailedCount}", arguments.InstanceMetadataList.Count, errors.Count);
+
+        return new UpdateInstanceResponse(updatedInstances.ToList(), errors.ToList());
```

Comment:
In theory you could cheat here by serializing as `IEnumerable<T>` and deserializing in the orchestration with `IReadOnlyList<T>` to remove the extra allocation/enumeration on serialization, but it's unnecessary (and would probably be messy to write).